snowplow-tracker/src/main/java/com/snowplowanalytics/core/tracker/TrackerControllerImpl.kt

```diff
@@ -70,6 +71,42 @@ class TrackerControllerImpl  // Constructors
         return tracker.track(event)
     }
 
+
+    override fun decorateLink(uri: Uri, parameters: List<CrossDeviceParameter>): Uri? {
+        // UserId is a required parameter of `_sp`
+        if (this.session?.userId == null) {
+            return null
+        }
+
+        val values = hashMapOf(
+            CrossDeviceParameter.SESSION_ID to (this.session?.sessionId ?: ""),
+            CrossDeviceParameter.SOURCE_ID to this.appId,
+            CrossDeviceParameter.SOURCE_PLATFORM to this.devicePlatform.value,
+            CrossDeviceParameter.USER_ID to (this.subject.userId ?: "")
```

Comment:
We will need to do a base64 encoding for the `userId`, `sourceId` and `reason` to make sure that they don't contain any dots which break the parsing on enrich.